File: rust/src/tools/send_video.rs
//! send_video — sends a video to the configured chat.

use frankenstein::{AsyncTelegramApi, ParseMode};
use frankenstein::input_file::FileUpload;
use frankenstein::methods::SendVideoParams;
use frankenstein::types::ReplyParameters;
use rmcp::model::CallToolResult;
use schemars::JsonSchema;
use serde::Deserialize;

use crate::telegram::{call_api, frank_to_tool_result, get_api, make_chat_id, resolve_chat, to_error, to_result, validate_caption};

#[derive(Debug, Deserialize, JsonSchema)]
pub struct SendVideoToolParams {
    /// Local file path, public HTTPS URL, or Telegram file_id
    pub video: String,
    pub caption: Option<String>,
    #[serde(default = "default_parse_mode")]
    pub parse_mode: String,
    pub duration: Option<u32>,
    pub width: Option<u32>,
    pub height: Option<u32>,
    pub disable_notification: Option<bool>,
    pub reply_to_message_id: Option<i32>,
}

fn default_parse_mode() -> String { "Markdown".to_owned() }

pub async fn impl_send_video(params: SendVideoToolParams) -> CallToolResult {
    let chat_id = match resolve_chat() {
        Ok(id) => id,
        Err(e) => return to_error(&e),
    };
    if let Some(ref cap) = params.caption {
        if let Some(err) = validate_caption(cap) { return to_error(&err); }
    }

    let video = FileUpload::String(params.video);
    let final_mode = if params.parse_mode == "HTML" { ParseMode::Html } else { ParseMode::MarkdownV2 };

    let mut video_params = SendVideoParams::builder()
        .chat_id(make_chat_id(&chat_id))
        .video(video)
        .parse_mode(final_mode)
        .build();

    if let Some(cap) = params.caption {
        let cap = if params.parse_mode == "Markdown" { crate::markdown::markdown_to_v2(&cap) } else { cap };
        video_params.caption = Some(cap);
    }
    if let Some(d) = params.duration { video_params.duration = Some(d); }
    if let Some(w) = params.width { video_params.width = Some(w); }
    if let Some(h) = params.height { video_params.height = Some(h); }
    if params.disable_notification == Some(true) { video_params.disable_notification = Some(true); }
    if let Some(reply_id) = params.reply_to_message_id {
        video_params.reply_parameters = Some(ReplyParameters::builder().message_id(reply_id).build());
    }

    match call_api(|| get_api().send_video(&video_params)).await {
        Ok(resp) => to_result(&serde_json::json!({ "message_id": resp.result.message_id })),
        Err(e) => frank_to_tool_result(e),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_parse_mode_is_markdown() {
        assert_eq!(default_parse_mode(), "Markdown");
    }
}

File: src/tools/send_video.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { getApi, toResult, toError, validateCaption, resolveChat, callApi, resolveMediaSource } from "../telegram.js";
import { resolveParseMode } from "../markdown.js";
import { cancelTyping, showTyping } from "../typing-state.js";
import { clearPendingTemp } from "../temp-message.js";
import { recordBotMessage } from "../session-recording.js";

export function register(server: McpServer) {
  server.registerTool(
    "send_video",
    {
      description: "Sends a video to the Telegram chat. Accepts a local file path, a public HTTPS URL, or a Telegram file_id. Supports optional caption, duration, and dimensions.",
      inputSchema: {
        video: z
        .string()
        .describe("Local absolute file path (e.g. /tmp/clip.mp4), a public HTTPS URL, or a Telegram file_id"),
      caption: z
        .string()
        .optional()
        .describe("Optional caption (up to 1024 chars)"),
      parse_mode: z
        .enum(["Markdown", "HTML", "MarkdownV2"])
        .default("Markdown")
        .describe("Markdown = standard Markdown auto-converted (default); MarkdownV2 = raw; HTML = HTML tags"),
      duration: z
        .number()
        .int()
        .optional()
        .describe("Video duration in seconds"),
      width: z
        .number()
        .int()
        .optional()
        .describe("Video width in pixels"),
      height: z
        .number()
        .int()
        .optional()
        .describe("Video height in pixels"),
      disable_notification: z
        .boolean()
        .optional()
        .describe("Send silently"),
      reply_to_message_id: z
        .number()
        .int()
        .optional()
        .describe("Reply to this message ID"),
      },
    },
    async ({ video, caption, parse_mode, duration, width, height, disable_notification, reply_to_message_id }) => {
      const chatId = resolveChat();
      if (typeof chatId !== "number") return toError(chatId);

      if (caption) {
        const capErr = validateCaption(caption);
        if (capErr) return toError(capErr);
      }

      const resolved = caption
        ? resolveParseMode(caption, parse_mode)
        : { text: undefined, parse_mode: undefined };

      const videoResult = resolveMediaSource(video);
      if ("code" in videoResult) return toError(videoResult);
      const videoSource = videoResult.source;

      await clearPendingTemp();
      try {
        await showTyping(120, "upload_video");
        const msg = await callApi(() =>
          getApi().sendVideo(chatId, videoSource, {
            caption: resolved.text,
            parse_mode: resolved.parse_mode,
            duration,
            width,
            height,
            disable_notification,
            reply_parameters: reply_to_message_id
              ? { message_id: reply_to_message_id }
              : undefined,
          })
        );
        cancelTyping();
        recordBotMessage({ content_type: "video", caption, message_id: msg.message_id });
        return toResult({
          message_id: msg.message_id,
          file_id: msg.video?.file_id,
          file_name: msg.video?.file_name,
          mime_type: msg.video?.mime_type,
          file_size: msg.video?.file_size,
          width: msg.video?.width,
          height: msg.video?.height,
          duration: msg.video?.duration,
        });
      } catch (err) {
        cancelTyping();
        return toError(err);
      }
    }
  );
}
